File: lightbulb-calendars/src/main/java/com/github/RooneyAndShadows/lightbulb/calendars/month/YearPageView.java
package com.github.rooneyandshadows.lightbulb.calendars.month;

import android.content.Context;
import android.util.AttributeSet;


import com.github.rooneyandshadows.commons.date.DateUtils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import static com.github.rooneyandshadows.lightbulb.calendars.month.MonthsAdapter.*;

class YearPageView extends RecyclerView {
    private MonthCalendarView calendarView;
    private int year;

    public YearPageView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        setSaveEnabled(true);
    }

    public YearPageView(int year, MonthCalendarView calendarView) {
        super(calendarView.getContext());
        this.year = year;
        this.calendarView = calendarView;
        setItemAnimator(null);
        setAdapter(new MonthsAdapter(calendarView, generateMonthsForYear()));
        setOverScrollMode(OVER_SCROLL_NEVER);
        setLayoutManager(new GridLayoutManager(getContext(), 4, VERTICAL, false));
    }

    public void select(int month) {
        getAdapter().selectMonth(month);
    }

    public void clearSelection() {
        getAdapter().clearSelection();
    }

    @Nullable
    @Override
    public MonthsAdapter getAdapter() {
        return (MonthsAdapter) super.getAdapter();
    }

    private ArrayList<MonthItem> generateMonthsForYear() {
        ArrayList<MonthItem> items = new ArrayList<>();
        for (Date date : DateUtils.getAllMonthsForYear(year)) {
            int currentYear = DateUtils.extractYearFromDate(date);
            int currentMonth = DateUtils.extractMonthOfYearFromDate(date);
            int[] currentMonthAsArray = new int[]{currentYear, currentMonth};
            int[] calendarSelectionAsArray = calendarView.getSelectionAsArray();
            boolean enabled = calendarView.isMonthEnabled(currentMonthAsArray[0], currentMonthAsArray[1]);
            boolean selected = Arrays.equals(calendarSelectionAsArray, currentMonthAsArray);
            items.add(new MonthItem(currentYear, currentMonth, selected, enabled));
        }
        return items;
    }
}

File: lightbulb-calendars/src/main/java/com/github/rooneyandshadows/lightbulb/calendars/month/adapter/YearPageView.kt
package com.github.rooneyandshadows.lightbulb.calendars.month.adapter

import android.content.Context
import android.util.AttributeSet
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.github.rooneyandshadows.lightbulb.calendars.month.MonthCalendarView
import com.github.rooneyandshadows.lightbulb.calendars.month.adapter.MonthsAdapter.*
import com.github.rooneyandshadows.java.commons.date.DateUtils
import com.github.rooneyandshadows.java.commons.date.DateUtilsOffsetDate
import java.util.*

internal class YearPageView : RecyclerView {
    private lateinit var calendarView: MonthCalendarView
    private var year = 0

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        isSaveEnabled = true
    }

    constructor(year: Int, calendarView: MonthCalendarView) : super(calendarView.context) {
        this.year = year
        this.calendarView = calendarView
        itemAnimator = null
        adapter = MonthsAdapter(calendarView, generateMonthsForYear())
        overScrollMode = OVER_SCROLL_NEVER
        layoutManager = GridLayoutManager(context, 4, VERTICAL, false)
    }

    @Override
    override fun getAdapter(): MonthsAdapter? {
        return super.getAdapter() as MonthsAdapter?
    }

    fun select(month: Int) {
        adapter!!.selectMonth(month)
    }

    fun clearSelection() {
        adapter!!.clearSelection()
    }

    private fun generateMonthsForYear(): MutableList<MonthItem> {
        val items: MutableList<MonthItem> = mutableListOf()
        for (date in DateUtilsOffsetDate.getAllMonthsForYear(year)) {
            val currentMonth = MonthEntry.fromDate(date)
            val calendarMonth = calendarView.selection
            val isSelected = currentMonth.compare(calendarMonth)
            val isEnabled = calendarView.isMonthEnabled(currentMonth)
            items.apply {
                add(MonthItem(currentMonth, isSelected, isEnabled))
            }
        }
        return items
    }
}